File: backend/controller/authController.js
const routeSchema = require("../models/routeSchema");
const User = require("../models/User");
const bcrypt = require("bcryptjs");


exports.signUp = async function (req, res) {
  try {
    const { name, email, password } = req.body;
    const existUser = await User.findOne({ email });
    if (existUser) return res.status(400).json({ msg: "User already exists" });

    const hashedPassword = await bcrypt.hash(password, 10); 
    const user = await User.create({ name, email, password: hashedPassword });

    console.log("✅ User saved:", user);
    res.status(201).json({ user: { name: user.name, email: user.email } });
  } catch (err) {
    console.error("Signup Error:", err.message);
    res.status(500).json({ msg: "Server error" });
  }
};

exports.logIn = async function (req, res) {
  try {
    const { email, password } = req.body;
    console.log("Incoming login:", email);

    const user = await User.findOne({ email });
    if (!user) {
      console.log("No user found for email:", email);
      return res.status(400).json({ msg: "Invalid email" });
    }

    const isMatch = await bcrypt.compare(password, user.password); 
    if (!isMatch) {
      console.log("Invalid password for:", email);
      return res.status(400).json({ msg: "Invalid password" });
    }

    res.status(200).json({ user: { name: user.name, email: user.email } });
  } catch (err) {
    console.error("Login Error:", err.message);
    res.status(500).json({ msg: "Server error" });
  }
};

File: backend_and_ML/controller/auth_controller.py
from fastapi import APIRouter, HTTPException, status
from database import user_collection
from fastapi import HTTPException
import bcrypt

router = APIRouter()


@router.post("/signup", status_code=status.HTTP_201_CREATED)
async def sign_up(payload: dict):
    try:
        name = payload.get("name")
        email = payload.get("email")
        password = payload.get("password")

        exist_user = await user_collection.find_one({"email": email})
        if exist_user:
            return {"msg": "User already exists"}

        hashed_password = bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt(10))

        user = {
            "name": name,
            "email": email,
            "password": hashed_password
        }

        result = await user_collection.insert_one(user)
        user["_id"] = str(result.inserted_id)

        print("User saved:", user)

        return {
            "user": {
                "name": name,
                "email": email
            }
        }

    except Exception as err:
        print("Signup Error:", str(err))
        raise HTTPException(
            status_code=500,
            detail="Server error"
        )


@router.post("/login")
async def log_in(payload: dict):
    email = payload.get("email")
    password = payload.get("password")

    print("Incoming login:", email)

    user = await user_collection.find_one({"email": email})
    if not user:
        print("No user found for email:", email)
        raise HTTPException(status_code=400, detail="Invalid email")

    is_match = bcrypt.checkpw(
        password.encode("utf-8"),
        user["password"]
    )

    if not is_match:
        print("Invalid password for:", email)
        raise HTTPException(status_code=400, detail="Invalid password")

    return {
        "user": {
            "name": user["name"],
            "email": user["email"]
        }
    }